README.md
# WebBro


## Problem statement: 
The project aims to develop a Python-based web browser, named WebBro, to provide a lightweight, customizable, and user-friendly alternative for internet browsing. The goal is to create a browser that is efficient, secure, and supports modern web standards.

## End-users of the Project:
- General internet users
- Developers
- Individuals looking for a customizable and lightweight web browsing experience

## Key Functions:

### 1. URL Editor
Users can manually input and edit URLs in the address bar.
This feature allows users to type or modify website URLs directly in the address bar, providing flexibility in navigating to specific web pages.

### 2. Navigation Buttons (Back and Forward)
Basic navigation buttons for moving backward and forward through the user's browsing history.
Users can easily navigate through previously visited pages by clicking the back and forward buttons, providing a familiar and essential browsing experience.

### 3. User-Defined Customizable Home Button
Users can set a custom homepage that can be accessed by clicking on a customizable home button.
This feature allows users to define a specific website or a local file as their homepage. Clicking on the home button will then take them to this designated page.

### 4. Reload Button
A button to refresh or reload the current web page.
Users can reload the page with a single click, ensuring they have the latest version of the content.

### 5. Toggle Between Light and Dark Modes
Users can switch between light and dark color schemes for the browser interface.
This feature provides users with the option to choose a light or dark mode based on their preference, enhancing readability and reducing eye strain.

### 6. Password Manager
Securely store and manage user passwords for websites.
WebBro's password manager allows users to save and retrieve passwords for different websites, streamlining the login process while maintaining security.

### 7. Customizable Search Engines
Users can add, remove, and customize search engines for the browser's search bar.
WebBro allows users to tailor their search experience by adding and configuring various search engines. Users can choose their preferred search engine for quick searches.

### 8. Settings Window
Access to a settings window with information about the browser and various customization options.
The settings window provides users with control over the browser's behavior and appearance. It includes options such as preferences, privacy settings, and information about the browser version.

### 9. Bookmark
Allow users to save and organize their favorite websites for quick access.
Provide a "Bookmark" option that users can click to save the current page to their bookmark list. Include a bookmark manager to organize and view saved bookmarks.

### 10. History
Track and display a user's browsing history.
Offer a "History" feature that allows users to view a list of recently visited websites. Clicking on an item in the history opens that page.

## Work Distribution

#### Rishi (PES2UG23CS479)

- Coordinate the overall project
- Implement settings window
- Implement customization of search engines
- Implement Bookmarking
- Oversee the integration of different components.

#### Rishabh (PES2UG23CSxxx)

- Assist in History Tracking
- Implement Password Manager functionality.
- Assist in implementing settings window

#### Rishil (PES2UG23CSxxx)

- Assist in Bookmarking
- Implement History tracking
- Implement light and dark mode
- Assist in Password managemer functionality

#### Raghav (PES2UG23CSxxx)

- Implement Text Bar for URL Input and Displaying Web Page Content.
- Implementing Navigation Buttons and reload button.
- Implement Home page customization.


history
# total_hours_wasted_here = 53

import sys
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtWebEngineWidgets import *


class MainWindow(QMainWindow):
    def __init__(self):
        super(MainWindow, self).__init__()
        self.browser = QWebEngineView()
        self.browser.setUrl(QUrl('https://zen-mauve.vercel.app/'))
        self.setCentralWidget(self.browser)
        self.showMaximized()
        self.box_visible = False
        self.settings_box_visible = False
        self.mode = "Dark"
        navbar = QToolBar()
        self.addToolBar(navbar)
        self.l1 = []
        self.search_engine = "brave"
        self.history = []
        self.history_box_visible = False

        icon_size = QSize(40, 40)  # Adjust the width and height as needed
        navbar.setIconSize(icon_size)

        # Set the stylesheet for the taskbar to make text white
        navbar.setStyleSheet("color: white;")

        button_width = 30

        back_btn = QToolButton(self)
        back_btn.setText('Back')
        back_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        back_btn.clicked.connect(self.browser.back)
        back_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(back_btn)

        forward_btn = QToolButton(self)
        forward_btn.setText('Forward')
        forward_btn.setIcon(QIcon('forward_icon.png'))  # Replace with your actual icon
        forward_btn.clicked.connect(self.browser.forward)
        forward_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(forward_btn)

        reload_btn = QToolButton(self)
        reload_btn.setText('Reload')
        reload_btn.setIcon(QIcon('reload_icon.png'))  # Replace with your actual icon
        reload_btn.clicked.connect(self.browser.reload)
        reload_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(reload_btn)

        self.home_btn = QToolButton(self)
        self.home_btn.setText('Home')
        self.home_btn.setIcon(QIcon('home_icon.png'))  # Replace with your actual icon
        self.home_btn.clicked.connect(self.navigate_home)
        self.home_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(self.home_btn)

        light_btn = QToolButton(self)
        light_btn.setText('Home')
        light_btn.setIcon(QIcon('home_icon.png'))  # Replace with your actual icon
        light_btn.clicked.connect(lambda: self.light_mode(navbar))
        light_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(light_btn)

        bookmark_btn = QToolButton(self)
        bookmark_btn.setText('bookmark')
        bookmark_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        bookmark_btn.clicked.connect(self.set_visible)
        bookmark_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(bookmark_btn)

        self.bookmark_box_widget = QFrame(self)
        self.bookmark_box_widget.setGeometry(184, 22, 400, 600)
        self.bookmark_box_widget.setFrameShape(QFrame.Box)
        self.bookmark_box_widget.setFrameShadow(QFrame.Raised)
        self.bookmark_box_widget.setLineWidth(2)
        self.bookmark_box_widget.setVisible(self.box_visible)
        self.bookmark_box_widget.setStyleSheet("background-color: #404040; color: white;")
        self.bookmark_box_layout = QVBoxLayout(self.bookmark_box_widget)
        self.bookmark_box_widget.setLayout(self.bookmark_box_layout)

        self.url_bar = QLineEdit()
        self.url_bar.returnPressed.connect(self.navigate_to_url)
        navbar.addWidget(self.url_bar)

        self.browser.urlChanged.connect(self.update_url)

        history_btn = QToolButton(self)
        history_btn.setText('History')
        history_btn.setIcon(QIcon('history.png'))  #Replace with whatever history icon you want
        history_btn.clicked.connect(self.show_history)
        history_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(history_btn)

        self.history_box_widget = QFrame(self)
        self.history_box_widget.setGeometry(184, 22, 400, 600)
        self.history_box_widget.setFrameShape(QFrame.Box)
        self.history_box_widget.setFrameShadow(QFrame.Raised)
        self.history_box_widget.setLineWidth(2)
        self.history_box_widget.setVisible(self.history_box_visible)
        self.history_box_widget.setStyleSheet("background-color: #404040; color: white;")
        self.history_box_layout = QVBoxLayout(self.history_box_widget)
        self.history_box_widget.setLayout(self.history_box_layout)

        #    self.search_engine_bar = QLineEdit()
        #    self.search_engine_bar.setPlaceholderText("Enter the search engine you want to use ex - google ")
        #    self.search_engine_bar.returnPressed.connect(self.search_engine_select)
        #    navbar.addWidget(self.search_engine_bar)

        combo_box = QComboBox()
        combo_box.addItem('brave')
        combo_box.addItem('bing')
        combo_box.addItem('google')
        combo_box.addItem('yahoo')
        combo_box.activated.connect(self.on_combobox_activated)
        navbar.addWidget(combo_box)

        new_window_btn = QToolButton(self)
        new_window_btn.setText('New Window')
        new_window_btn.clicked.connect(self.open_new_window)
        navbar.addWidget(new_window_btn)

        settings_btn = QToolButton(self)
        settings_btn.setText('settings')
        settings_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        settings_btn.clicked.connect(self.set_settings_visible)
        settings_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(settings_btn)

        self.settings_widget = QFrame(self)
        self.settings_widget.setGeometry(184, 22, 400, 600)
        self.settings_widget.setFrameShape(QFrame.Box)
        self.settings_widget.setFrameShadow(QFrame.Raised)
        self.settings_widget.setLineWidth(2)

webbro1.py
import sys
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtWebEngineWidgets import *


class MainWindow(QMainWindow):
    def __init__(self):
        super(MainWindow, self).__init__()
        self.browser = QWebEngineView()
        self.browser.setUrl(QUrl('https://zen-mauve.vercel.app/'))
        self.setCentralWidget(self.browser)
        self.showMaximized()
        self.box_visible = False 

        navbar = QToolBar()
        self.addToolBar(navbar)

        icon_size = QSize(40,40)  # Adjust the width and height as needed
        navbar.setIconSize(icon_size)

        # Set the stylesheet for the taskbar to make text white
        navbar.setStyleSheet("color: white;")

        button_width = 30  

        back_btn = QToolButton(self)
        back_btn.setText('Back')
        back_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        back_btn.clicked.connect(self.browser.back)
        back_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(back_btn)

        forward_btn = QToolButton(self)
        forward_btn.setText('Forward')
        forward_btn.setIcon(QIcon('forward_icon.png'))  # Replace with your actual icon
        forward_btn.clicked.connect(self.browser.forward)
        forward_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(forward_btn)

        reload_btn = QToolButton(self)
        reload_btn.setText('Reload')
        reload_btn.setIcon(QIcon('reload_icon.png'))  # Replace with your actual icon
        reload_btn.clicked.connect(self.browser.reload)
        reload_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(reload_btn)

        home_btn = QToolButton(self)
        home_btn.setText('Home')
        home_btn.setIcon(QIcon('home_icon.png'))  # Replace with your actual icon
        home_btn.clicked.connect(self.navigate_home)
        home_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(home_btn)

        # Create the bookmark button
        self.bookmark_btn = QAction('Bookmark', self)
        self.bookmark_btn.triggered.connect(self.toggle_bookmark_box)
        navbar.addAction(self.bookmark_btn)

        # Create the history button
        self.history_btn = QAction('History', self)
        self.history_btn.triggered.connect(self.toggle_history_box)
        navbar.addAction(self.history_btn)

        # Create the bookmark box widget
        self.bookmark_box_widget = QFrame(self)
        self.bookmark_box_widget.setGeometry(184, 22, 400, 600)
        self.bookmark_box_widget.setFrameShape(QFrame.Box)
        self.bookmark_box_widget.setFrameShadow(QFrame.Raised)
        self.bookmark_box_widget.setLineWidth(2)
        self.bookmark_box_widget.setVisible(self.box_visible)
        self.bookmark_box_widget.setStyleSheet("background-color: #404040; color: #872341;")

        # Create the bookmark list widget inside the bookmark box widget
        self.bookmark_list_widget = QListWidget(self.bookmark_box_widget)
        self.bookmark_list_widget.itemDoubleClicked.connect(self.load_item)

        # Set the stylesheet for the bookmark list items
        self.bookmark_list_widget.setStyleSheet("color: #F2F2F2;")

        # Create a layout for the bookmark box widget
        bookmark_layout = QVBoxLayout(self.bookmark_box_widget)
        bookmark_layout.addWidget(self.bookmark_list_widget)

        # Create the history box widget
        self.history_box_widget = QFrame(self)
        self.history_box_widget.setGeometry(184, 22, 400, 600)
        self.history_box_widget.setFrameShape(QFrame.Box)
        self.history_box_widget.setFrameShadow(QFrame.Raised)
        self.history_box_widget.setLineWidth(2)
        self.history_box_widget.setVisible(self.box_visible)
        self.history_box_widget.setStyleSheet("background-color: #353535; color: white;")

        # Create the history list widget inside the history box widget
        self.history_list_widget = QListWidget(self.history_box_widget)
        self.history_list_widget.itemDoubleClicked.connect(self.load_item)

        # Set the stylesheet for the history list items
        self.history_list_widget.setStyleSheet("color: white;")

        # Create a layout for the history box widget
        history_layout = QVBoxLayout(self.history_box_widget)
        history_layout.addWidget(self.history_list_widget)

        # Initialize the current box widget to None
        self.current_box_widget = None

        # Create the URL bar
        self.url_bar = QLineEdit()
        self.url_bar.returnPressed.connect(self.navigate_to_url)
        navbar.addWidget(self.url_bar)

        # Connect the URL changed signal to the update_url method
        self.browser.urlChanged.connect(self.update_url)

        # Initialize the history list
        self.history = []

    def toggle_bookmark_box(self):
        self.toggle_box(self.bookmark_box_widget, self.bookmark_list_widget)

    def toggle_history_box(self):
        self.toggle_box(self.history_box_widget, self.history_list_widget)

    def toggle_box(self, box_widget, list_widget):
        self.box_visible = not self.box_visible

        if self.current_box_widget and self.current_box_widget != box_widget:
            self.current_box_widget.setVisible(False)

        box_widget.setVisible(self.box_visible)

        if self.box_visible:
            current_url = self.url_bar.text()
            self.add_item_to_list(list_widget, current_url)

        self.current_box_widget = box_widget

    def navigate_home(self):
        self.browser.setUrl(QUrl('https://zen-mauve.vercel.app/'))

    def add_item_to_list(self, list_widget, item):
        list_item = QListWidgetItem(item)
        list_widget.addItem(list_item)

    def load_item(self, item):
        url = item.text()
        self.url_bar.setText(url)
        self.browser.setUrl(QUrl(url))

    def navigate_to_url(self):
        url = self.url_bar.text()

        if "." in url:
            if "https://" in url or "http://" in url:
                self.browser.setUrl(QUrl(url))
            else:
                self.browser.setUrl(QUrl("https://" + url))
        else:
            temp = url.split(" ")
            url = "https://search.brave.com/search?q="
            for i in temp:
                url = url + "+" + i
            url = url + "&source=web"
            self.browser.setUrl(QUrl(url))

    def update_url(self, q):
        url = q.toString()
        self.url_bar.setText(url)
        self.history.append(url)
        self.add_item_to_list(self.history_list_widget, url)

app = QApplication(sys.argv)
QApplication.setApplicationName('My Cool Browser')
window = MainWindow()
# Set the background color of the taskbar to black
window.setStyleSheet("background-color: black;")
app.exec_()


webbro_simpler.py
# Dear programmer:
# when I wrote this code,
# only god and I knew how it worked.
# Now, only god knows it!
#
# Therefore, if you are trying to optimize 
# this routine and it fails ( most surely ),
# please increase this counter as a
# warning for the next person:
#
# total_hours_wasted_here = 53

import sys
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtWebEngineWidgets import *

class MainWindow(QMainWindow):
    def __init__(self):
        super(MainWindow, self).__init__()
        self.browser = QWebEngineView()
        self.browser.setUrl(QUrl('https://zen-mauve.vercel.app/'))
        self.setCentralWidget(self.browser)
        self.showMaximized()
        self.box_visible = False 
        self.settings_box_visible = False
        self.mode = "Dark"
        navbar = QToolBar()
        self.addToolBar(navbar)
        self.l1 = []
        self.search_engine = "brave"


        icon_size = QSize(40,40)  # Adjust the width and height as needed
        navbar.setIconSize(icon_size)

        # Set the stylesheet for the taskbar to make text white
        navbar.setStyleSheet("color: white;")

        button_width = 30  

        back_btn = QToolButton(self)
        back_btn.setText('Back')
        back_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        back_btn.clicked.connect(self.browser.back)
        back_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(back_btn)

        forward_btn = QToolButton(self)
        forward_btn.setText('Forward')
        forward_btn.setIcon(QIcon('forward_icon.png'))  # Replace with your actual icon
        forward_btn.clicked.connect(self.browser.forward)
        forward_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(forward_btn)


        reload_btn = QToolButton(self)
        reload_btn.setText('Reload')
        reload_btn.setIcon(QIcon('reload_icon.png'))  # Replace with your actual icon
        reload_btn.clicked.connect(self.browser.reload)
        reload_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(reload_btn)

        self.home_btn = QToolButton(self)
        self.home_btn.setText('Home')
        self.home_btn.setIcon(QIcon('home_icon.png'))  # Replace with your actual icon
        self.home_btn.clicked.connect(self.navigate_home)
        self.home_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(self.home_btn)


        light_btn = QToolButton(self)
        light_btn.setText('Home')
        light_btn.setIcon(QIcon('home_icon.png'))  # Replace with your actual icon
        light_btn.clicked.connect(lambda: self.light_mode(navbar))
        light_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(light_btn)

        bookmark_btn = QToolButton(self)
        bookmark_btn.setText('bookmark')
        bookmark_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        bookmark_btn.clicked.connect(self.set_visible)
        bookmark_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(bookmark_btn)



        self.bookmark_box_widget = QFrame(self)
        self.bookmark_box_widget.setGeometry(184, 22, 400, 600)
        self.bookmark_box_widget.setFrameShape(QFrame.Box)
        self.bookmark_box_widget.setFrameShadow(QFrame.Raised)
        self.bookmark_box_widget.setLineWidth(2)
        self.bookmark_box_widget.setVisible(self.box_visible)
        self.bookmark_box_widget.setStyleSheet("background-color: #404040; color: white;")
        self.bookmark_box_layout = QVBoxLayout(self.bookmark_box_widget)
        self.bookmark_box_widget.setLayout(self.bookmark_box_layout)

        self.url_bar = QLineEdit()
        self.url_bar.returnPressed.connect(self.navigate_to_url)
        navbar.addWidget(self.url_bar)

        self.browser.urlChanged.connect(self.update_url)

    #    self.search_engine_bar = QLineEdit()
    #    self.search_engine_bar.setPlaceholderText("Enter the search engine you want to use ex - google ")
    #    self.search_engine_bar.returnPressed.connect(self.search_engine_select)
    #    navbar.addWidget(self.search_engine_bar)

        combo_box = QComboBox()
        combo_box.addItem('brave')
        combo_box.addItem('bing')
        combo_box.addItem('google')
        combo_box.addItem('yahoo')        
        combo_box.activated.connect(self.on_combobox_activated)
        navbar.addWidget(combo_box)

        new_window_btn = QToolButton(self)
        new_window_btn.setText('New Window')
        new_window_btn.clicked.connect(self.open_new_window)
        navbar.addWidget(new_window_btn)

        settings_btn = QToolButton(self)
        settings_btn.setText('settings')
        settings_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        settings_btn.clicked.connect(self.set_settings_visible)
        settings_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(settings_btn)



        self.settings_widget = QFrame(self)
        self.settings_widget.setGeometry(184, 22, 400, 600)
        self.settings_widget.setFrameShape(QFrame.Box)
        self.settings_widget.setFrameShadow(QFrame.Raised)
        self.settings_widget.setLineWidth(2)
        self.settings_widget.setVisible(self.settings_box_visible)
        self.settings_widget.setStyleSheet("background-color: #404040; color: white;")
        self.settings_layout = QVBoxLayout(self.settings_widget)
        self.settings_widget.setLayout(self.settings_layout)

        self.show_home_btn_checkbox = QCheckBox('Enable Feature')
        self.settings_layout.addWidget(self.show_home_btn_checkbox)
        self.show_home_btn_checkbox.stateChanged.connect(self.some_method)

        self.home_btn.setVisible(not self.show_home_btn_checkbox.isChecked())



    def some_method(self):
        self.home_btn.setVisible(not self.show_home_btn_checkbox.isChecked())




    def set_settings_visible(self):
        self.settings_box_visible = not self.settings_box_visible
        
        self.settings_widget.setVisible(self.settings_box_visible)

    def open_new_window(self):
        new_window = MainWindow()
        new_window.show()

    def open_settings(self):
        settings_window = SettingsWindow(self)
        settings_window.exec_()

    def on_combobox_activated(self):
        self.search_engine = self.sender().currentText()

    #def search_engine_select(self):
    #   self.search_engine =  self.search_engine_bar.text()

    def set_visible(self):
        self.box_visible = not self.box_visible
        if self.box_visible:
            current_url = self.url_bar.text()
            if current_url not in self.l1:
                self.l1.append(current_url)
                label = QLabel(current_url)
                self.bookmark_box_layout.addWidget(label)
        self.bookmark_box_widget.setVisible(self.box_visible)


    def light_mode(self, navbar):
        if self.mode == "Dark":
            navbar.setStyleSheet("color: black;")
            self.setStyleSheet("background-color: white;")
            self.mode="Light"
        elif self.mode == "Light":
            navbar.setStyleSheet("color: white;")
            self.setStyleSheet("background-color: black;")
            self.mode="Dark"
        

    def navigate_home(self):
        self.browser.setUrl(QUrl('https://zen-mauve.vercel.app/'))

    def navigate_to_url(self):
        url = self.url_bar.text()

        if "." in url:
            if "https://" in url or "http://" in url:
                self.browser.setUrl(QUrl(url))
            else:
                self.browser.setUrl(QUrl("https://" + url))
        else:
            temp = url.split(" ")
            url = "https://search."+self.search_engine+".com/search?q="
            for i in temp:
                url = url + "+" + i
            url = url + "&source=web"
            self.browser.setUrl(QUrl(url))

    def update_url(self, q):
        url = q.toString()
        self.url_bar.setText(url)

app = QApplication(sys.argv)
QApplication.setApplicationName('My Cool Browser')
window = MainWindow()
# Set the background color of the taskbar to black
window.setStyleSheet("background-color: black;")
app.exec_()



webbro_simple.py
# Dear programmer:
# when i wrote this code,
# only god and i knew how it worked.
# Now, only god knows it!
#
# Therefore, if you are trying to optimize this routine
# and it fails ( most surely ),
# please increase this counter as a
# warning for the next person:
#
# total_hours_wasted_here = 69

import sys
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtWebEngineWidgets import *


class MainWindow(QMainWindow):
    def __init__(self):
        super(MainWindow, self).__init__()
        self.browser = QWebEngineView()
        self.browser.setUrl(QUrl('https://zen-mauve.vercel.app/'))
        self.setCentralWidget(self.browser)
        self.showMaximized()
        self.box_visible = False
        self.settings_box_visible = False
        self.mode = "Dark"
        navbar = QToolBar()
        self.addToolBar(navbar)
        self.l1 = []
        self.search_engine = "brave"
        self.history = []
        self.history_box_visible = False

        icon_size = QSize(40, 40)  # Adjust the width and height as needed
        navbar.setIconSize(icon_size)

        # Set the stylesheet for the taskbar to make text white
        navbar.setStyleSheet("color: white;")

        button_width = 30

        back_btn = QToolButton(self)
        back_btn.setText('Back')
        back_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        back_btn.clicked.connect(self.browser.back)
        back_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(back_btn)

        forward_btn = QToolButton(self)
        forward_btn.setText('Forward')
        forward_btn.setIcon(QIcon('forward_icon.png'))  # Replace with your actual icon
        forward_btn.clicked.connect(self.browser.forward)
        forward_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(forward_btn)

        reload_btn = QToolButton(self)
        reload_btn.setText('Reload')
        reload_btn.setIcon(QIcon('reload_icon.png'))  # Replace with your actual icon
        reload_btn.clicked.connect(self.browser.reload)
        reload_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(reload_btn)

        self.home_btn = QToolButton(self)
        self.home_btn.setText('Home')
        self.home_btn.setIcon(QIcon('home_icon.png'))  # Replace with your actual icon
        self.home_btn.clicked.connect(self.navigate_home)
        self.home_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(self.home_btn)

        light_btn = QToolButton(self)
        light_btn.setText('Home')
        light_btn.setIcon(QIcon('home_icon.png'))  # Replace with your actual icon
        light_btn.clicked.connect(lambda: self.light_mode(navbar))
        light_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(light_btn)

        bookmark_btn = QToolButton(self)
        bookmark_btn.setText('bookmark')
        bookmark_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        bookmark_btn.clicked.connect(self.set_visible)
        bookmark_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(bookmark_btn)

        self.bookmark_box_widget = QFrame(self)
        self.bookmark_box_widget.setGeometry(184, 22, 400, 600)
        self.bookmark_box_widget.setFrameShape(QFrame.Box)
        self.bookmark_box_widget.setFrameShadow(QFrame.Raised)
        self.bookmark_box_widget.setLineWidth(2)
        self.bookmark_box_widget.setVisible(self.box_visible)
        self.bookmark_box_widget.setStyleSheet("background-color: #404040; color: white;")
        self.bookmark_box_layout = QVBoxLayout(self.bookmark_box_widget)
        self.bookmark_box_widget.setLayout(self.bookmark_box_layout)

        self.url_bar = QLineEdit()
        self.url_bar.returnPressed.connect(self.navigate_to_url)
        navbar.addWidget(self.url_bar)

        self.browser.urlChanged.connect(self.update_url)

        history_btn = QToolButton(self)
        history_btn.setText('History')
        history_btn.setIcon(QIcon('history.png'))  #Replace with whatever history icon you want
        history_btn.clicked.connect(self.show_history)
        history_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(history_btn)

        self.history_box_widget = QFrame(self)
        self.history_box_widget.setGeometry(184, 22, 400, 600)
        self.history_box_widget.setFrameShape(QFrame.Box)
        self.history_box_widget.setFrameShadow(QFrame.Raised)
        self.history_box_widget.setLineWidth(2)
        self.history_box_widget.setVisible(self.history_box_visible)
        self.history_box_widget.setStyleSheet("background-color: #404040; color: white;")
        self.history_box_layout = QVBoxLayout(self.history_box_widget)
        self.history_box_widget.setLayout(self.history_box_layout)

        #    self.search_engine_bar = QLineEdit()
        #    self.search_engine_bar.setPlaceholderText("Enter the search engine you want to use ex - google ")
        #    self.search_engine_bar.returnPressed.connect(self.search_engine_select)
        #    navbar.addWidget(self.search_engine_bar)

        combo_box = QComboBox()
        combo_box.addItem('brave')
        combo_box.addItem('bing')
        combo_box.addItem('google')
        combo_box.addItem('yahoo')
        combo_box.activated.connect(self.on_combobox_activated)
        navbar.addWidget(combo_box)

        new_window_btn = QToolButton(self)
        new_window_btn.setText('New Window')
        new_window_btn.clicked.connect(self.open_new_window)
        navbar.addWidget(new_window_btn)

        settings_btn = QToolButton(self)
        settings_btn.setText('settings')
        settings_btn.setIcon(QIcon('back_icon.png'))  # Replace with your actual icon
        settings_btn.clicked.connect(self.set_settings_visible)
        settings_btn.setFixedSize(button_width, button_width)
        navbar.addWidget(settings_btn)

        self.settings_widget = QFrame(self)
        self.settings_widget.setGeometry(184, 22, 400, 600)
        self.settings_widget.setFrameShape(QFrame.Box)
        self.settings_widget.setFrameShadow(QFrame.Raised)
        self.settings_widget.setLineWidth(2)
        self.settings_widget.setVisible(self.settings_box_visible)
        self.settings_widget.setStyleSheet("background-color: #404040; color: white;")
        self.settings_layout = QVBoxLayout(self.settings_widget)
        self.settings_widget.setLayout(self.settings_layout)

        self.show_home_btn_checkbox = QCheckBox('Enable Feature')
        self.settings_layout.addWidget(self.show_home_btn_checkbox)
        self.show_home_btn_checkbox.stateChanged.connect(self.some_method)

        self.home_btn.setVisible(not self.show_home_btn_checkbox.isChecked())

    def some_method(self):
        self.home_btn.setVisible(not self.show_home_btn_checkbox.isChecked())

    def set_settings_visible(self):
        self.settings_box_visible = not self.settings_box_visible

        self.settings_widget.setVisible(self.settings_box_visible)

    def open_new_window(self):
        new_window = MainWindow()
        new_window.show()

    def open_settings(self):
        settings_window = SettingsWindow(self)
        settings_window.exec_()

    def on_combobox_activated(self):
        self.search_engine = self.sender().currentText()

    # def search_engine_select(self):
    #   self.search_engine =  self.search_engine_bar.text()

    def set_visible(self):
        self.box_visible = not self.box_visible
        if self.box_visible:
            current_url = self.url_bar.text()
            if current_url not in self.l1:
                self.l1.append(current_url)
                label = QLabel(current_url)
                self.bookmark_box_layout.addWidget(label)
        self.bookmark_box_widget.setVisible(self.box_visible)

    def light_mode(self, navbar):
        if self.mode == "Dark":
            navbar.setStyleSheet("color: black;")
            self.setStyleSheet("background-color: white;")
            self.mode = "Light"
        elif self.mode == "Light":
            navbar.setStyleSheet("color: white;")
            self.setStyleSheet("background-color: black;")
            self.mode = "Dark"

    def navigate_home(self):
        self.browser.setUrl(QUrl('https://zen-mauve.vercel.app/'))

    def navigate_to_url(self):
        url = self.url_bar.text()

        if "." in url:
            if "https://" in url or "http://" in url:
                self.browser.setUrl(QUrl(url))
            else:
                self.browser.setUrl(QUrl("https://" + url))
        else:
            temp = url.split(" ")
            url = "https://search." + self.search_engine + ".com/search?q="
            for i in temp:
                url = url + "+" + i
            url = url + "&source=web"
            self.browser.setUrl(QUrl(url))
        

    def update_url(self, q):
        url = q.toString()
        print(url)
        self.url_bar.setText(url)
        self.history.append(url)
        

    # def show_history(self):
    #     self.settings_box_visible = not self.settings_box_visible
    #     history_text = "\n".join(self.history)
    #     msg_box = QMessageBox()
    #     msg_box.setText(history_text)
    #     msg_box.setStyleSheet("QMessageBox { background-color: lightblue; } QLabel { color: darkblue; }")
    #     msg_box.setWindowTitle("Browsing History")
    #     msg_box.exec_()
    #     self.history_box_widget.setVisible(self.history_box_visible)

    def show_history(self):
        self.history_box_visible = not self.history_box_visible
        history_text = "\n".join(self.history)

        while self.history_box_layout.count():
            item = self.history_box_layout.takeAt(0)
            widget = item.widget()
            if widget:
                widget.setParent(None)
        # for item in range(0,len(self.history),2):
        #     label = QLabel(self.history[item], self)
        #     self.history_box_layout.addWidget(label)
        for item in self.history:
             label = QLabel(item, self)
             self.history_box_layout.addWidget(label)
        self.history_box_widget.setVisible(self.history_box_visible)




app = QApplication(sys.argv)
QApplication.setApplicationName('My Cool Browser')
window = MainWindow()
# Set the background color of the taskbar to black
window.setStyleSheet("background-color: black;")
app.exec_()
